ports/stm32/boards/i2splltable.py

```diff
@@ -0,0 +1,220 @@
+    i2splltable = []
+    for bits in (16, 32):
+        for rate in (8_000, 11_025, 12_000, 16_000, 22_050, 24_000, 32_000, 44_100, 48_000):
+            i2spll_candidates = []
+            for plli2sn in mcu.range_plli2sn:
+                for plli2sr in mcu.range_plli2sr:
+                    I2SxCLK = hse // pllm * plli2sn // plli2sr
+                    if I2SxCLK < 192_000_000:
+                        # compute I2S clock generator parameters: i2sdiv, odd
+                        tmp = ((((I2SxCLK // (bits * 2)) * 10) // rate)) + 5
+                        tmp = tmp // 10
+                        odd = tmp & 1
+                        i2sdiv = (tmp - odd) // 2
+                        Fs = I2SxCLK / ((bits * 2) * ((2 * i2sdiv) + odd))
+                        error = (abs(Fs - rate) / rate) * 100
+                        i2spll_candidates.append(
+                            i2spll(
+                                bits=bits,
+                                rate=rate,
+                                plli2sn=plli2sn,
+                                plli2sr=plli2sr,
+                                i2sdiv=i2sdiv,
+                                odd=odd,
+                                error=error,
+                            )
+                        )
+            # sort based on error
+            i2spll_candidates_sorted = sorted(i2spll_candidates, key=lambda x: x.error)
+            # select the best candidate
+            i2splltable.append(i2spll_candidates_sorted[0])
+    return i2splltable
+
+
+def generate_c_table(i2splltable, hse, pllm):
+    print("// MAKE generated file, created by i2splltable.py: DO NOT EDIT")
+    print("// This table is used in machine_i2s.c")
+    print(f"// HSE_VALUE = {hse}")
+    print(f"// MICROPY_HW_CLK_PLLM = {pllm} \n")
+    print("#define I2S_PLL_CONFIG \\")
+    print("{ \\")
+
+    for i2spll in i2splltable:
+        print(
+            f"    {{{i2spll.rate}, "
+            f"{i2spll.bits}, "
+            f"{i2spll.plli2sr}, "
+            f"{i2spll.plli2sn} }}, "
+            f"/* i2sdiv: {int(i2spll.i2sdiv)}, "
+            f"odd: {i2spll.odd}, "
+            f"rate error % (desired vs actual)%: {i2spll.error:.4f} */ \\"
+        )
+    print("}")
+
+
+def search_header_for_hse_value(filename, val):
+    regex_inc = re.compile(r'#include "(boards/[A-Za-z0-9_./]+)"')
+    regex_def = re.compile(r"#define +(HSE_VALUE) +\((\(uint32_t\))?([0-9]+)\)")
+    with open(filename) as f:
+        for line in f:
+            line = line.strip()
+            m = regex_inc.match(line)
+            if m:
+                # Search included file
+                search_header_for_hse_value(m.group(1), val)
+                continue
+            m = regex_def.match(line)
+            if m:
+                # Found HSE_VALUE
+                if m.group(1) == "HSE_VALUE":
+                    val[0] = int(m.group(3))
+    return val
+
+
+def search_header_for_pllm(filename, val):
```

Comment:
This function is structurally the same as the one above.  I suggest factoring them by providing a generic `search_header` function that takes the regex's and does the search.  Then this and the above function become simple one-line functions that call `search_header` with the appropriate regex's.

Reply:
A generic function was created as suggested.